sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala

```diff
@@ -231,3 +229,30 @@ object GroupingID {
     if (SQLConf.get.integerGroupingIdEnabled) IntegerType else LongType
   }
 }
+
+object GroupingAnalytics {
+  def unapply(exprs: Seq[Expression])
+  : Option[(Seq[Seq[Expression]], Seq[Seq[Expression]], Seq[Expression])] = {
+    if (!exprs.exists(_.find(_.isInstanceOf[BaseGroupingSets]).isDefined)) {
+      None
+    } else {
+      val groups = exprs.flatMap {
+        case gs: BaseGroupingSets => gs.groupByExprs
+        case others: Expression => others :: Nil
+      }
+      val groupingSets = exprs.flatMap {
+        case gs: BaseGroupingSets => gs.groupingSets
+        case _ => Nil
+      }
+      val unmergedSelectedGroupByExprs = exprs.map {
+        case gs: BaseGroupingSets => gs.selectedGroupByExprs
+        case others: Expression => Seq(Seq(others))
+      }
+      val selectedGroupByExprs = unmergedSelectedGroupByExprs.init
+        .foldLeft(unmergedSelectedGroupByExprs.last) { (x, y) =>
+          for (a <- x; b <- y) yield a ++ b
+        }
+      Some(selectedGroupByExprs, groupingSets, groups.distinct)
```

Comment:
let's follow `BaseGroupingSets.groupByExprs`, use semanticEquals to dedup instead of `distinct `

Reply:
where do we use the second return parameter?

Reply:
> where do we use the second return parameter?

Oh, we don't need it after your last pr. remove it.

Reply:
> let's follow `BaseGroupingSets.groupByExprs`, use semanticEquals to dedup instead of `distinct `

Done